File: src/test/java/net/sf/rails/game/state/GenericStateTest.java
package net.sf.rails.game.state;

import static org.fest.assertions.api.Assertions.assertThat;
import static org.junit.Assert.*;

import net.sf.rails.game.state.GenericState;
import net.sf.rails.game.state.Item;
import net.sf.rails.game.state.Root;

import org.junit.Before;
import org.junit.Test;

public class GenericStateTest {

    private final static String DEFAULT_ID = "Default";
    private final static String INIT_ID = "Init"; 
    private final static String ITEM_ID = "Item";
    private final static String ANOTHER_ID = "Another";

    private Root root;
    
    private GenericState<Item> stateDefault;
    private GenericState<Item> stateInit;
    
    private Item item, another_item;

    @Before
    public void setUp() {
        root = StateTestUtils.setUpRoot();
        
        
        item = new AbstractItemImpl(root, ITEM_ID);
        another_item = new AbstractItemImpl(root, ANOTHER_ID);
        
        stateDefault = GenericState.create(root, DEFAULT_ID);
        stateInit = GenericState.create(root, INIT_ID, item);
    }
    
    @Test
    public void testValue() {
        assertNull(stateDefault.value());
        assertSame(item, stateInit.value());
    }
    
    @Test
    public void testSet() {
        stateDefault.set(item);
        assertSame(item, stateDefault.value());
        stateDefault.set(null);
        assertNull(stateDefault.value());
        stateInit.set(another_item);
        assertSame(another_item, stateInit.value());
    }
    
    @Test
    public void testSetSameIgnored() {
        stateDefault.set(null);
        stateInit.set(item);
        StateTestUtils.close(root);
        assertThat(StateTestUtils.getPreviousChangeSet(root).getStates()).doesNotContain(stateDefault, stateInit);
    }

    @Test
    public void testUndoRedo() {
        assertNull(stateDefault.value());
        assertSame(item, stateInit.value());

        stateDefault.set(item);
        stateInit.set(another_item);
        assertSame(item, stateDefault.value());
        assertSame(another_item, stateInit.value());

        StateTestUtils.close(root);
        // remark: stateInit is an internal (isObservable = false)
        assertThat(StateTestUtils.getPreviousChangeSet(root).getStates()).contains(stateDefault);
        
        StateTestUtils.undo(root);
        assertNull(stateDefault.value());
        assertSame(item, stateInit.value());

        StateTestUtils.redo(root);
        assertSame(item, stateDefault.value());
        assertSame(another_item, stateInit.value());
    }

}

File: dotnet/GameStateTest/GenericStateTest.cs
using GameLib.Net.Game.State;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GameStateTest
{
    [TestClass]
    public class GenericStateTest
    {
        private const string DEFAULT_ID = "Default";
        private const string INIT_ID = "Init";
        private const string ITEM_ID = "Item";
        private const string ANOTHER_ID = "Another";

        private Root root;

        private GenericState<IItem> stateDefault;
        private GenericState<IItem> stateInit;

        private IItem item, another_item;

        [TestInitialize]
        public void SetUp()
        {
            root = StateTestUtils.SetUpRoot();


            item = new AbstractItemImpl(root, ITEM_ID);
            another_item = new AbstractItemImpl(root, ANOTHER_ID);

            stateDefault = GenericState<IItem>.Create(root, DEFAULT_ID);
            stateInit = GenericState<IItem>.Create(root, INIT_ID, item);
        }

        [TestMethod]
        public void TestValue()
        {
            Assert.IsNull(stateDefault.Value);
            Assert.AreSame(item, stateInit.Value);
        }

        [TestMethod]
        public void TestSet()
        {
            stateDefault.Set(item);
            Assert.AreSame(item, stateDefault.Value);
            stateDefault.Set(null);
            Assert.IsNull(stateDefault.Value);
            stateInit.Set(another_item);
            Assert.AreSame(another_item, stateInit.Value);
        }

        [TestMethod]
        public void TestSetSameIgnored()
        {
            stateDefault.Set(null);
            stateInit.Set(item);
            StateTestUtils.Close(root);
            var a = StateTestUtils.GetPreviousChangeSet(root).GetStates();
            Assert.IsTrue(!(a.Contains(stateDefault) || a.Contains(stateInit)));
            //assertThat(StateTestUtils.getPreviousChangeSet(root).getStates()).doesNotContain(stateDefault, stateInit);
        }

        [TestMethod]
        public void TestUndoRedo()
        {
            Assert.IsNull(stateDefault.Value);
            Assert.AreSame(item, stateInit.Value);

            stateDefault.Set(item);
            stateInit.Set(another_item);
            Assert.AreSame(item, stateDefault.Value);
            Assert.AreSame(another_item, stateInit.Value);

            StateTestUtils.Close(root);
            // remark: stateInit is an internal (isObservable = false)
            Assert.IsTrue(StateTestUtils.GetPreviousChangeSet(root).GetStates().Contains(stateDefault));
            //assertThat(StateTestUtils.getPreviousChangeSet(root).getStates()).contains(stateDefault);

            StateTestUtils.Undo(root);
            Assert.IsNull(stateDefault.Value);
            Assert.AreSame(item, stateInit.Value);

            StateTestUtils.Redo(root);
            Assert.AreSame(item, stateDefault.Value);
            Assert.AreSame(another_item, stateInit.Value);
        }

    }
}
